rocket_sqlite_oracle/src/main.rs

```diff
@@ -103,148 +111,99 @@ async fn price_at_time(mut conn: Connection<ExampleDB>, timestamp: i64) -> Json<
+    let row = sqlx::query!(
+        r#"
+        with above as (
+          select *
+          from signatures
+          where timestamp >= ?
+          order by timestamp
+          limit 1
+        ),
 
-async fn load_secret_key_from_file(path: impl AsRef<Path>) -> Result<secp256k1::SecretKey> {
-    let path = path.as_ref();
-    let bytes = tokio::fs::read(path)
-        .await
-        .with_context(|| format!("No secret file at {}", path.display()))?;
-    let secret_key = secp256k1::SecretKey::from_slice(&bytes)?;
+        below as (
+          select *
+          from signatures
+          where timestamp < ?
+          order by timestamp desc
+          limit 1
+        ),
 
-    Ok(secret_key)
-}
+        opts as (
+          select * from above
+          union all
+          select * from below
+        )
 
-async fn write_secret_key_to_file(secret_key: &secp256k1::SecretKey, path: PathBuf) -> Result<()> {
-    if let Some(parent) = path.parent() {
-        tokio::fs::DirBuilder::new()
-            .recursive(true)
-            .create(parent)
-            .await
-            .with_context(|| {
-                format!(
-                    "Could not create directory for secret file: {}",
-                    parent.display()
-                )
-            })?;
-    }
-    let mut file = tokio::fs::OpenOptions::new()
-        .write(true)
-        .create_new(true)
-        .open(&path)
-        .await
-        .with_context(|| format!("Could not generate secret file at {}", path.display()))?;
-
-    file.write_all(secret_key.as_ref()).await?;
+        select
+          timestamp,
+          bid_price as price,
+          bid_signature as signature
+        from opts
+        order by abs(? - timestamp)
+        limit 1;
+        "#,
+        timestamp,
+        timestamp,
+        timestamp
+    )
+    .fetch_one(&mut *conn)
+    .await
+    .unwrap();
+    let res = db::USDTickerResponse {
+        timestamp: row.timestamp.try_into().unwrap(),
+        price: row.price.try_into().unwrap(),
+        signature: Signature::from_compact(&row.signature).unwrap(),
+    };
 
-    Ok(())
+    Json(res)
 }
 
 #[rocket::main]
 async fn main() -> Result<(), anyhow::Error> {
-    let cli = Cli::from_args();
+    let xcli = cli::Cli::from_args();
 
     // TODO
     // this logic-flow isn't terribly user-friendly
-    let oracle_secret_key = match cli.generate_oracle_secret {
+    let oracle_secret = match xcli.generate_secret {
         true => {
-            let secret_key = secp256k1::SecretKey::new(&mut thread_rng());
-            write_secret_key_to_file(&secret_key, cli.oracle_secret).await?;
+            let skey = SecretKey::new(&mut thread_rng());
+            xcli.write_secret_key_to_file(&skey).await?;
 
-            secret_key
+            skey
         }
-        false => load_secret_key_from_file(&cli.oracle_secret).await?,
+        false => xcli.load_secret_key_from_file(&xcli.secret).await?,
     };
```

Comment:
If you want `write_secret_key_to_file` and `load_secret_key_from_file` to be functions on `Cli`, then this entire block can move into a single function on `Cli` because all the information is present internally.

Here for example:

```rust
xcli.load_secret_key_from_file(&xcli.secret)
```

`xcli` is already passed to `load_secret_key_from_file` via the `self` parameter. No need to pass `secret` in separately, it is already available via `self.secret`.

Also, the decision on logic based on `write_secret_key_to_file` vs `load_secret_key_from_file` is based on `xcli.generate_secret`.

This kind of interaction is quite "chatty" and leads to weird indirections.

Either condense it down to a single function call on `Cli` or don't make `write_secret_key_to_file` and `load_secret_key_from_file` functions on `Cli`.